core/src/main/java/org/elasticsearch/common/path/PathTrie.java

```diff
@@ -220,10 +349,59 @@ public T retrieve(String path, Map<String, String> params) {
             return rootValue;
         }
         int index = 0;
-        // supports initial delimiter.
+
+        // Supports initial delimiter.
         if (strings.length > 0 && strings[0].isEmpty()) {
             index = 1;
         }
-        return root.retrieve(strings, index, params);
+
+        return root.retrieve(strings, index, params, trieMatchingMode);
+    }
+
+    /**
+     * Returns an iterator of the objects stored in the {@code PathTrie}, using
+     * all possible {@code TrieMatchingMode} modes. The {@code paramSupplier}
+     * is called between each invocation of {@code next()} to supply a new map
+     * of parameters.
+     */
+    public Iterator<T> retrieveAll(String path, Supplier<Map<String, String>> paramSupplier) {
+        return new PathTrieIterator<>(this, path, paramSupplier);
+    }
+
+    class PathTrieIterator<T> implements Iterator<T> {
+
+        private final List<TrieMatchingMode> modes = new ArrayList<>(4);
+        private final Supplier<Map<String, String>> paramSupplier;
+        private final PathTrie<T> trie;
+        private final String path;
+
+        PathTrieIterator(PathTrie trie, String path, Supplier<Map<String, String>> paramSupplier) {
+            this.path = path;
+            this.trie = trie;
+            this.paramSupplier = paramSupplier;
+            modes.add(TrieMatchingMode.EXPLICIT_NODES_ONLY);
```

Comment:
Oh, but I see what you're saying, you're just saying for contructing it, sure I'll change that